### PR TITLE
8332488: Add JVMTI DataDumpRequest to the debug agent

### PR DESCRIPTION
**Pull request by plummercj**
JVMTI has a somewhat unique event called DataDumpRequest. One way it is triggered is via the JVMTI.data_dump jcmd, which causes JVMTI to send the DataDumpRequest event to all agents that have registered for the event callback. The agent is free to do pretty much what it wants during the callback, but the normal usage is to dump anything that might be useful for debugging the agent. In the case of the debug agent, it could dump internal data like the list of known threads and event handlers. After ranked monitor support is complete, it can also dump the state of all jvmti raw monitors that the debug agent uses.

I decided to not enable this feature by default, and not make public the option to enable it. This should only be used by developers working on the debug agent, or by users when requested to do so (by debug agent developers) to help debug a debug agent problem.

Most of the code executed during the data dump was only available for debug builds, so I've made it available for all builds. Their addition does not affect product builds except for adding a small footprint.

TBD is directing the output to a file. This is useful for some of the debugger tests that don't include the debuggee output in the log. This seems to be the case for most com/sun/jdi tests. I decided not to include it for this first pass since it is rather disruptive and detracts from the main changes being made.

testing tbd: run all tier1, tier2, and. tie5 svc tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332488](https://bugs.openjdk.org/browse/JDK-8332488): Add JVMTI DataDumpRequest to the debug agent (**Enhancement** - P4)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20367/head:pull/20367` \
`$ git checkout pull/20367`

Update a local copy of the PR: \
`$ git checkout pull/20367` \
`$ git pull https://git.openjdk.org/jdk.git pull/20367/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20367`

View PR using the GUI difftool: \
`$ git pr show -t 20367`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20367.diff">https://git.openjdk.org/jdk/pull/20367.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20367#issuecomment-2253712092)

### CODE REVIEW

**Review by sspitsyn — COMMENTED** [review]
First, it is a good idea to implement/enable this.
It looks good in general. Need another pass to understand the implications better.
Did you consider to add a test for this?

**Comment by plummercj** [comment]
>  Did you consider to add a test for this?
Good idea. I'll add one.

File: test/jdk/com/sun/jdi/DataDumpTest.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+import com.sun.jdi.request.ClassPrepareRequest;
+/**
+ * @test
+ * @bug  8332488
+ * @summary Unit test for testing debug agent support for JVMTI.data_dump jcmd.
+ *
+ * @library /test/lib
+ * @modules jdk.jdi
+ * @build DataDumpTest
+ * @run main/othervm/timeout=15 DataDumpTest
```
Comment (lmesnik) [inline test/jdk/com/sun/jdi/DataDumpTest.java:52]: Is the othervm is really needed in this test? 
Also, it is unclear why the timeout=15 and not usual 120?

**Comment by plummercj** [comment]
This PR is ready for review again. I added a test case and also added a rather long comment documenting the support for data dumps.

**Review by lmesnik — APPROVED** [review]
In general changes looks good, there are only a couple small test questions.

**Comment by plummercj** [comment]
FYI, this is what the (entire) debuggee output looks like for this test:

```
Deuggee output:
Listening for transport dt_socket at address: 42509
Debug Agent Data Dump
=== START DUMP ===
suspendAllCount: 0
Dumping runningThreads:
Thread: node = 0x149684000f10, jthread = 0x1497504da382
	name: Notification Thread
	suspendCount: 0
Thread: node = 0x14969c000f10, jthread = 0x1497504da352
	name: Common-Cleaner
	suspendCount: 0
Thread: node = 0x149750583a60, jthread = 0x1497504da332
	name: Attach Listener
	suspendCount: 0
Thread: node = 0x1497505835c0, jthread = 0x1497504da32a
	name: Signal Dispatcher
	suspendCount: 0
Thread: node = 0x149750583120, jthread = 0x1497504da322
	name: Finalizer
	suspendCount: 0
Thread: node = 0x149750582c50, jthread = 0x1497504da31a
	name: Reference Handler
	suspendCount: 0
Thread: node = 0x1497505827b0, jthread = 0x1497504da312
	name: main
	suspendCount: 1

Dumping runningVThreads:

Dumping otherThreads:

Handlers for EI_CLASS_PREPARE(7)
node(0x1496a40034e0) handlerID(4) suspendPolicy(1) permanent(0)
ClassMatch: classPattern(DataDumpTestTarg)
node(0x1496a4002d80) handlerID(2) suspendPolicy(0) permanent(0)

Handlers for EI_CLASS_UNLOAD(8)
node(0x1496a4002de0) handlerID(3) suspendPolicy(0) permanent(0)

Handlers for EI_VM_DEATH(20)
node(0x1496a4004a20) handlerID(0) suspendPolicy(0) permanent(1)
=== END DUMP ===
Debuggee started
```

**Comment by lmesnik** [comment]
> FYI, this is what the (entire) debuggee output looks like for this test:
> 
> ```
> Deuggee output:
>

Should we see the same report for CTRL+\ or after sending SIGQUIT signal? I see it should cause this request.

**Comment by plummercj** [comment]
> > FYI, this is what the (entire) debuggee output looks like for this test:
> > ```
> > Deuggee output:
> > ```
> 
> Should we see the same report for CTRL+\ or after sending SIGQUIT signal? I see it should cause this request.

Yes, I believe so. In fact during testing I think more than once I saw the dump just by issuing a random jcmd. I believe what happened is the debug agent had suspended all threads so that the attach would time out (due to Signal Handler thread being suspended). The failure_handler would timeout a number of jcmds in sequence. I think this causes a number of SIGQUIT signals to queue up. Once the debugger eventually resumed all threads, the first SIGQUIT would cause the attach listener to start up, and subsequent ones caused a thread dump and data dump. Kind of strange, but doesn't really have anything to with my changes, and I've seen thread dumps like this before due to attach issues.

**Comment by plummercj** [comment]
@sspitsyn and @lmesnik Can I get final reviews from both of you? Thanks.

File: test/jdk/com/sun/jdi/DataDumpTest.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+import com.sun.jdi.request.EventRequest;
+import com.sun.jdi.request.ClassPrepareRequest;
+/**
+ * @test
+ * @bug  8332488
+ * @summary Unit test for testing debug agent support for JVMTI.data_dump jcmd.
+ *
+ * @library /test/lib
+ * @modules jdk.jdi
+ * @build DataDumpTest
```
[the remark this answers is not in the document]
Comment (lmesnik) [reply test/jdk/com/sun/jdi/DataDumpTest.java:51]: ough, I mean 'is @build DataDumpTest' needed to build DataDumpTestTarg? The DataDumpTest should be built without explicit @build command.

[the remark this answers is not in the document]
Comment (plummercj) [reply test/jdk/com/sun/jdi/DataDumpTest.java]: Ok. I removed the `@build`

File: test/jdk/com/sun/jdi/DataDumpTest.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+import com.sun.jdi.VMDisconnectedException;
+import com.sun.jdi.connect.AttachingConnector;
+import com.sun.jdi.connect.Connector;
+import com.sun.jdi.connect.IllegalConnectorArgumentsException;
+import com.sun.jdi.event.ClassPrepareEvent;
+import com.sun.jdi.event.Event;
+import com.sun.jdi.event.EventSet;
+import com.sun.jdi.request.EventRequest;
+import com.sun.jdi.request.ClassPrepareRequest;
+/**
```
Comment (sspitsyn) [inline test/jdk/com/sun/jdi/DataDumpTest.java:44]: Nit: We normally have an empty line here.

File: src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c [hunk 4]
```diff
@@ -1330,6 +1330,44 @@ cbVMDeath(jvmtiEnv *jvmti_env, JNIEnv *env)
     LOG_MISC(("END cbVMDeath"));
 }
 
+/**
+ * Event callback for JVMTI_EVENT_DATA_DUMP_REQUEST
+ *
+ * This callback is made when a JVMTI data dump is requested. The common way of doing
+ * this is with "jcmd <pid> JVMTI.data_dump".
```
Comment (sspitsyn) [inline src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:1337]: It is nice to have this here.

**Review by sspitsyn — APPROVED** [review]
Looks good.

**Review by lmesnik — APPROVED** [review]

**Comment by plummercj** [comment]
@sspitsyn or @lmesnik I need a final review since I made a minor comment change. Thanks.

**Review by lmesnik — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]

**Comment by plummercj** [comment]
Thanks Leonid and Serguei!

/integrate

**Pull request closed without merging**